services/supervisor/src/transactions.rs

```diff
@@ -89,6 +120,125 @@ impl ValidateInput for StartService {
+        // To prevent multiple service change proposition in one request
+        let mut service_ids = BTreeSet::new();
+
+        // Perform config verification.
+        for change in &propose.changes {
+            match change {
+                ConfigChange::Consensus(config) => {
+                    if multiple_consensus_propose {
+                        trace!("Discarded multiple consensus change propositions in one request.");
+                        return Err(Error::MalformedConfigPropose.into());
+                    }
+                    multiple_consensus_propose = true;
+
+                    config
+                        .validate()
+                        .map_err(|e| (Error::MalformedConfigPropose, e))?;
+                }
+
+                ConfigChange::Service(config) => {
+                    if service_ids.contains(&config.instance_id) {
+                        trace!("Discarded multiple service change propositions in one request.");
+                        return Err(Error::MalformedConfigPropose.into());
+                    }
+                    service_ids.insert(config.instance_id);
+
+                    context
+                        .interface::<ConfigureCall>(config.instance_id)
+                        .verify_config(config.params.clone())
+                        .map_err(|e| (Error::MalformedConfigPropose, e))?;
+                }
+            }
+        }
+
+        let propose_hash = propose.object_hash();
+        config_confirms.confirm(&propose_hash, author);
+
+        let config_entry = ConfigProposeEntry {
+            config_propose: propose,
+            propose_hash,
+        };
+        schema.config_propose_with_hash_entry().set(config_entry);
+
+        Ok(())
+    }
+
+    fn confirm_config_change(
+        &self,
+        context: TransactionContext,
+        vote: ConfigVote,
+    ) -> Result<(), ExecutionError> {
+        let blockchain_height = blockchain::Schema::new(context.fork()).height();
+        let schema = Schema::new(context.instance.name, context.fork());
+
+        // Verifies that transaction author is validator.
+        let author = context.caller().author().ok_or(Error::UnknownAuthor)?;
+        let mut config_confirms = schema.config_confirms();
+        config_confirms
+            .validator_id(author)
+            .ok_or(Error::UnknownAuthor)?;
+
+        // Verifies that this config proposal is registered.
+        if let Some(entry) = schema.config_propose_with_hash_entry().get() {
+            if entry.propose_hash != vote.propose_hash {
+                return Err(Error::ConfigProposeNotRegistered.into());
+            }
+
+            let config_propose = entry.config_propose;
+            // Verifies that we doesn't reach deadline height.
+            if config_propose.actual_from <= blockchain_height {
+                return Err(Error::DeadlineExceeded.into());
+            }
+
+            if config_confirms.confirmed_by(&entry.propose_hash, &author) {
```

Comment:
This piece of code looks very strange, may you explain what it happens here or look carefully to error type.